lib/charms/prometheus_pushgateway_k8s/v0/pushgateway.py

```diff
@@ -0,0 +1,172 @@
+
+logger = logging.getLogger(__name__)
+
+
+# The unique Charmhub library identifier, never change it
+LIBID = "a0065690fe484ef296d5847fcbf1d728"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# the key in the relation data
+RELATION_KEY = "push-endpoint"
+
+
+class PrometheusPushgatewayProvider(Object):
+    """Provider side for the Prometheus Pushgateway.
+
+    This class is to be used by the Prometheus Pushgateway charm, please use
+    the PrometheusPushgatewayRequirer class if you're bulding a charm and want to
+    use this library to integrate with the Prometheus Pushgateway.
+    """
+
+    def __init__(self, charm: CharmBase, relation_name: str, port: int):
+        super().__init__(charm, relation_name)
+        self.port = port
+        self.app = charm.app
+        events = charm.on[relation_name]
+        self.framework.observe(events.relation_created, self._on_relation_changed)
+        self.framework.observe(events.relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: RelationEvent):
+        """Send the push endpoint info."""
+        relation_data = event.relation.data[self.app]
+        relation_data[RELATION_KEY] = json.dumps(
+            {
+                "hostname": socket.getfqdn(),
+                "port": self.port,
+            }
+        )
+
+
+class PrometheusPushgatewayRequirer(Object):
+    """Requirer side for the Prometheus Pushgateway."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Construct the interface for the Prometheus Pushgateway.
+
+        Args:
+            charm: a `CharmBase` object that manages this object. Typically,
+                this is `self` in the instantiating class.
+            relation_name: the name of the relation (whatever was used
+                in the `requires` section in `metadata.yaml` for
+                the `pushgateway` interface.
+        """
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+
+    @property
+    def _pushgateway_url(self) -> Optional[str]:
+        """Build the pushgateway url using the relation data (if present, else return None)."""
+        relation = self.model.get_relation(self._relation_name)
+        if relation is None:
+            logger.warning(
+                "Prometheus Pushgateway Requirer not ready: "
+                "charm not related to the Pushgateway service")
+            return None
+        raw_data = relation.data[relation.app].get(RELATION_KEY)
+        if raw_data is None:
+            logger.warning(
+                "Prometheus Pushgateway Requirer not ready: still no data in the relation")
+            return None
+        data = json.loads(raw_data)
+        return "http://{hostname}:{port}/".format_map(data)
```

Comment:
Should we have the relation data hold the full URL instead of splitting it in the provider and recombining in the requirer?

Reply:
I like this idea... not because splitting and recombining (because it's really never splitted, but combined only once), but because it also transmits HTTP or HTTPS.

Reply:
Done.